### PR TITLE
Attach GameLogic to Blocks multiple times

### PR DESCRIPTION
**Pull request by Tyrsis**
This allows multiple scripts to attach to blocks without collisions.  If a block already has gamelogic attached, it will keep track of new gamelogic attachments and run those as well.

This also allows scripts to attach only to subtypes that they care about, so their logic only runs on the blocks they care about by SubTypeName.  I modified MyEntityComponentDescriptor to allow you to specify multiple subtypenames, so you can attach the same gamelogic to multiple subtypes (small and large for example).

Should be completely backwards compatible.  If new attributes are not used, it defaults to old behavior.  This also enhances old behavior in that multiple mods won't collide anymore if they want to attach logic to the same block as all scripts gamelogic will run.

Example:
`[MyEntityComponentDescriptor(typeof(MyObjectBuilder_Beacon), "SmallBlockRadar", "LargeBlockRadar")]`

When the above attribute is attached to a class that modifies beacon behavior, it only attaches to entities that have the subtypename of SmallBlockRadar or LargeBlockRadar.

### CODE REVIEW

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs [hunk 1]
```diff
@@ -477,18 +504,33 @@ public void CreateSync()
         public virtual void UpdateOnceBeforeFrame()
         {
             m_gameLogic.UpdateOnceBeforeFrame();
+			if (m_gameLogicSet != null)
```
Comment (harag-on-steam) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: This looks like inconsistent indentation settings.

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs [hunk 2]
```diff
@@ -100,6 +100,13 @@ public MyGameLogicComponent GameLogic
             set { Components.Add<MyGameLogicComponent>(value); }
         }
 
+		private HashSet<MyGameLogicComponent> m_gameLogicSet = null;
```
Comment (harag-on-steam) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: Why not just make m_gameLogic a set? Now the class has two fields that basically serve the same purpose.

Comment (Tyrsis) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: I felt there was no need to needlessly enumerate when gamelogic had nothing attached to it.

**Comment by THDigi** [comment]
Great, thanks!

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs (hunk 2 above)
Comment (harag-on-steam) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: Hm, that could also be achieved by introducing a MyCompoundGameLogicComponent that has the collection and would be used as soon as there are more than one MyGameLogicComponents for the same entity.

Or maybe it would be enough to just introduce 3 delegates for the three Update*() methods and attach the set of MyGameLogicComponents to them? That could also serve as a more fine-grained way for update-notifications. If a component isn't interested in Update*100 it isn't registered to that delegate (by some opt-out or opt-in mechanism).

Now that I look at it again, is HashSet appropriate here? How is equality of the elements determined?

File: .gitignore [hunk 3]
```diff
@@ -38,3 +38,4 @@ user.props
 3rd/SteamSDK/release/x64
 3rd/VRage.Native/debug/x64
 3rd/VRage.Native/release/x64
+global.props
```
Comment (THDigi) [inline .gitignore]: You should use user.props instead and undo this change :P

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs (hunk 2 above)
Comment (Tyrsis) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: I thought about delegates, but then felt that would just be inconsistent with how logic is currently attached, and a bit confusing, while also forcing mods to change (a bad thing).

File: .gitignore (hunk 3 above)
Comment (Tyrsis) [inline .gitignore]: Yeah, woops, let me try to fix things up again, heh.

**Comment by Tyrsis** [comment]
Sorry about all the changes, I need to learn to git properly.

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs [hunk 4]
```diff
@@ -100,6 +100,13 @@ public MyGameLogicComponent GameLogic
             set { Components.Add<MyGameLogicComponent>(value); }
```
Comment (harag-on-steam) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: `Components.Add` does not happen for the elements of `m_gameLogicSet`. That can't be a good thing.

Looking at `Components.Add` I guess each logic component should be added with it's own type or the components will overwrite each other in the component-list.

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs [hunk 5]
```diff
@@ -477,17 +504,32 @@ public void CreateSync()
         public virtual void UpdateOnceBeforeFrame()
         {
             m_gameLogic.UpdateOnceBeforeFrame();
```
Comment (harag-on-steam) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: Beside the `Update*` methods there's also Init, MarkForClose, Close and UpdatingStopped that need to be handled.

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs [hunk 6]
```diff
@@ -463,6 +470,26 @@ void Components_ComponentRemoved(Type t, Sandbox.Common.Components.MyComponentBa
...
         }
 
+        public void AssignGamelogicFromHashSet(HashSet<Type> items)
+        {
+            bool first = true;
+            foreach (Type logicType in items)
+            {
+                if (first)
+                {
+                    GameLogic = (MyGameLogicComponent)Activator.CreateInstance(logicType);
```
Comment (harag-on-steam) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: Hm, iteration order is unpredictable so it's unknown which component will end up in here.
Also, this single component can still be overwritten by setting GameLogic.

**Comment by harag-on-steam** [comment]
And please don't take this the wrong way. I think this is an important change so I'm putting a lot of thought into the corner cases :)

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs (hunk 6 above)
Comment (Tyrsis) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: Order does not matter here.  This is just to fill GameLogic first before filling the set.

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs (hunk 5 above)
Comment (Tyrsis) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: Yes good eye, I will fix those.  UpdatingStopped wasn't ever called originally.  So I didn't include it.  Not sure if that was on purpose or not.

File: Sources/Sandbox.Game/Game/Entities/MyEntity.cs (hunk 4 above)
Comment (Tyrsis) [inline Sources/Sandbox.Game/Game/Entities/MyEntity.cs]: I'm not actually sure what they are doing with components.  I almost feel like it's a Type container for validation or something.  When I did my tests before, it didn't really seem to matter.  I will ask someone in Keen though.

**Comment by Tyrsis** [comment]
There is an issue still with this I'm tracking down.

**Comment by Tyrsis** [comment]
This should be fine now.  All tests passed.  Multiple attachments to a block type and sub block type, all working together in harmony.

**Comment by THDigi** [comment]
This'll work even if multiple mods attach to the same typeid, right ?

**Comment by Tyrsis** [comment]
Yes, this works with multiple mods attaching to same type, and subtype, and any combination within.

**Comment by THDigi** [comment]
Ok so I understood it correctly then, awesome job :+1:

File: Sources/Sandbox.Common/Components/MyComponentBase.cs [hunk 7]
```diff
@@ -11,17 +11,19 @@ namespace Sandbox.Common.Components
...
         {
             EntityBuilderType = entityBuilderType;
+            EntityBuilderSubTypeNames = entityBuilderSubTypeNames;
         }
     }
 
     public abstract class MyComponentBase
     {
         public IMyEntity Entity { get { return CurrentContainer != null ? CurrentContainer.Entity : null; } }  // to be obsolete once components are finished
-        public MyComponentContainer CurrentContainer { get; private set; }
```
Comment (OndrejPetrzilka) [inline Sources/Sandbox.Common/Components/MyComponentBase.cs]: This seems wrong, protected is not enough?

**Comment by Phoenix84** [comment]
Sadly it's not completely backwards compatible.
If there is only one script to attach, it all works fine. If there are multiple, however, Entity.GameLogic no longer represents what it's assumed to.

For example, in my Stargate mod, I case entity.GameLogic to my own class to access properties and methods on another entity.
I can't do that anymore. I looked at the code, but I can't find a way to get a specific MyGameLogicComponent object out of it.

I can filter by subtype, but that's not a solution for me, as I have a second mod that adds to it, but also includes the same script to be used independently if desired.
